# Notebook 21 — Per-Cap Frequency Capping

**Date**: 2026-02-27

**Context**: `db:postgresQuery` = 43.8% de toutes les traces (964/2205).
Le modèle GRU overfit sur cette cap dominante au détriment des 330 autres.

**Objectif**: Mesurer l'impact du per-cap capping (max N exemples/target)
sur la distribution des données d'entraînement et la qualité des prédictions.

**Méthode**: Farthest-point sampling sur les intent embeddings pour
garder les intents les plus diversifiés par cap.

In [1]:
import psycopg2
import numpy as np
import json
import os
from collections import Counter, defaultdict

conn = psycopg2.connect(os.environ.get('DATABASE_URL', 'postgres://casys:Kx9mP2vL7nQ4wRzT@localhost:5432/casys'))
cur = conn.cursor()
print('Connected')

Connected


## 1. Load training data distribution

In [2]:
# Load all traces with cap target + intent embedding
cur.execute("""
    SELECT et.id,
           cr.namespace || ':' || cr.action as cap_name,
           et.intent_embedding
    FROM execution_trace et
    JOIN workflow_pattern wp ON et.capability_id = wp.pattern_id
    JOIN capability_records cr ON cr.workflow_pattern_id = wp.pattern_id
    WHERE et.intent_embedding IS NOT NULL
      AND et.executed_path IS NOT NULL
      AND array_length(et.executed_path, 1) > 0
    ORDER BY et.executed_at DESC
""")
traces = cur.fetchall()
print(f'{len(traces)} traces loaded')

# Count per cap
cap_counts = Counter()
cap_intents = defaultdict(list)  # cap_name → list of intent embeddings
for trace_id, cap_name, intent_emb in traces:
    cap_counts[cap_name] += 1
    emb = np.array(json.loads(intent_emb) if isinstance(intent_emb, str) else intent_emb, dtype=np.float32)
    cap_intents[cap_name].append(emb)

print(f'{len(cap_counts)} unique caps')
print(f'\nTop 10 caps:')
for cap, cnt in cap_counts.most_common(10):
    pct = 100 * cnt / len(traces)
    print(f'  {cap:40s} {cnt:5d} ({pct:5.1f}%)')

# Distribution stats
counts = np.array(list(cap_counts.values()))
print(f'\nDistribution: median={np.median(counts):.0f}, mean={np.mean(counts):.1f}, p90={np.percentile(counts,90):.0f}, p99={np.percentile(counts,99):.0f}, max={np.max(counts)}')
print(f'Gini coefficient: {1 - 2 * np.sum(np.sort(counts).cumsum()) / (len(counts) * np.sum(counts)):.3f}')

2163 traces loaded


331 unique caps

Top 10 caps:
  db:postgresQuery                           934 ( 43.2%)
  db:queryLatestTrace                         67 (  3.1%)
  db:tableSchemas                             47 (  2.2%)
  fs:readJson                                 39 (  1.8%)
  system:systemctl                            33 (  1.5%)
  cap:rename                                  33 (  1.5%)
  syson:addSensorToTcs                        25 (  1.2%)
  erpnext:stockChart                          24 (  1.1%)
  infra:listDockerContainers                  23 (  1.1%)
  docker:listContainers                       23 (  1.1%)

Distribution: median=2, mean=6.5, p90=8, p99=37, max=934
Gini coefficient: 0.754


## 2. Farthest-point sampling (replicate TypeScript logic)

In [3]:
def farthest_point_sample(embeddings, k, seed=42):
    """Greedy farthest-point sampling. Returns indices of selected points."""
    n = len(embeddings)
    if n <= k:
        return list(range(n))
    
    rng = np.random.RandomState(seed)
    selected = [rng.randint(n)]
    min_dist = np.full(n, np.inf)
    
    # Update distances from first pivot
    diffs = embeddings - embeddings[selected[0]]
    min_dist = np.sum(diffs * diffs, axis=1)
    min_dist[selected[0]] = 0
    
    for _ in range(1, k):
        best_idx = np.argmax(min_dist)
        if min_dist[best_idx] <= 0:
            break
        selected.append(best_idx)
        min_dist[best_idx] = 0
        
        diffs = embeddings - embeddings[best_idx]
        new_dist = np.sum(diffs * diffs, axis=1)
        min_dist = np.minimum(min_dist, new_dist)
    
    return selected


def cap_examples_per_target(cap_intents, max_per_cap, seed=42):
    """Apply per-cap frequency capping with farthest-point sampling."""
    result = {}
    stats = {'capped': 0, 'total_before': 0, 'total_after': 0}
    
    for cap_name, intents in cap_intents.items():
        n = len(intents)
        stats['total_before'] += n
        
        if n <= max_per_cap:
            result[cap_name] = intents
            stats['total_after'] += n
        else:
            emb_matrix = np.stack(intents)
            selected_idx = farthest_point_sample(emb_matrix, max_per_cap, seed)
            result[cap_name] = [intents[i] for i in selected_idx]
            stats['total_after'] += len(selected_idx)
            stats['capped'] += 1
    
    return result, stats

# Quick test
test_embs = [np.random.randn(8).astype(np.float32) for _ in range(100)]
idx = farthest_point_sample(np.stack(test_embs), 10)
print(f'FPS test: {len(test_embs)} points → {len(idx)} selected')
print('OK')

FPS test: 100 points → 10 selected
OK


## 3. Compare capping thresholds

In [4]:
print(f'{"Max/cap":>8s} | {"Total":>6s} | {"Dropped":>8s} | {"Top cap %":>9s} | {"Unique caps":>11s} | {"Gini":>6s}')
print('-' * 65)

for max_cap in [999999, 100, 50, 30, 20, 10]:
    capped, stats = cap_examples_per_target(cap_intents, max_cap)
    
    capped_counts = {k: len(v) for k, v in capped.items()}
    total = sum(capped_counts.values())
    top_cap_pct = 100 * max(capped_counts.values()) / total if total > 0 else 0
    vals = np.array(list(capped_counts.values()))
    gini = 1 - 2 * np.sum(np.sort(vals).cumsum()) / (len(vals) * np.sum(vals))
    dropped = stats['total_before'] - stats['total_after']
    dropped_pct = 100 * dropped / stats['total_before']
    
    label = 'NONE' if max_cap > 99999 else str(max_cap)
    print(f'{label:>8s} | {total:6d} | {dropped:5d} ({dropped_pct:4.1f}%) | {top_cap_pct:8.1f}% | {len(capped_counts):11d} | {gini:6.3f}')

 Max/cap |  Total |  Dropped | Top cap % | Unique caps |   Gini
-----------------------------------------------------------------
    NONE |   2163 |     0 ( 0.0%) |     43.2% |         331 |  0.754
     100 |   1329 |   834 (38.6%) |      7.5% |         331 |  0.604
      50 |   1262 |   901 (41.7%) |      4.0% |         331 |  0.583
      30 |   1135 |  1028 (47.5%) |      2.6% |         331 |  0.544
      20 |   1039 |  1124 (52.0%) |      1.9% |         331 |  0.508
      10 |    900 |  1263 (58.4%) |      1.1% |         331 |  0.443


## 4. Intent diversity analysis: FPS vs random

In [5]:
# For the dominant cap, compare FPS vs random sampling intent diversity
dominant_cap = cap_counts.most_common(1)[0][0]
dominant_intents = cap_intents[dominant_cap]
emb_matrix = np.stack(dominant_intents)

# Normalize for cosine
norms = np.linalg.norm(emb_matrix, axis=1, keepdims=True)
emb_normed = emb_matrix / np.maximum(norms, 1e-12)

print(f'Dominant cap: {dominant_cap} ({len(dominant_intents)} intents)')

for max_k in [10, 20, 30, 50]:
    # FPS sampling
    fps_idx = farthest_point_sample(emb_matrix, max_k)
    fps_embs = emb_normed[fps_idx]
    fps_sim = fps_embs @ fps_embs.T
    fps_mean_sim = (fps_sim.sum() - np.trace(fps_sim)) / (len(fps_idx) * (len(fps_idx) - 1)) if len(fps_idx) > 1 else 0
    
    # Random sampling
    rng = np.random.RandomState(42)
    rand_idx = rng.choice(len(dominant_intents), size=min(max_k, len(dominant_intents)), replace=False)
    rand_embs = emb_normed[rand_idx]
    rand_sim = rand_embs @ rand_embs.T
    rand_mean_sim = (rand_sim.sum() - np.trace(rand_sim)) / (len(rand_idx) * (len(rand_idx) - 1)) if len(rand_idx) > 1 else 0
    
    print(f'  k={max_k:3d}: FPS mean pairwise sim={fps_mean_sim:.4f}  Random={rand_mean_sim:.4f}  (lower=more diverse)')

Dominant cap: db:postgresQuery (934 intents)
  k= 10: FPS mean pairwise sim=0.5734  Random=0.6299  (lower=more diverse)
  k= 20: FPS mean pairwise sim=0.6043  Random=0.6675  (lower=more diverse)
  k= 30: FPS mean pairwise sim=0.6117  Random=0.6693  (lower=more diverse)
  k= 50: FPS mean pairwise sim=0.6301  Random=0.6697  (lower=more diverse)


## 5. Simulated Hit@1 impact with capping

In [6]:
# Load cap SHGAT embeddings for scoring simulation
cur.execute("""
    SELECT DISTINCT ON (cr.namespace, cr.action)
        cr.namespace || ':' || cr.action as cap_name,
        COALESCE(wp.shgat_embedding, wp.intent_embedding) as embedding
    FROM workflow_pattern wp
    JOIN capability_records cr ON cr.workflow_pattern_id = wp.pattern_id
    WHERE wp.code_snippet IS NOT NULL
      AND wp.intent_embedding IS NOT NULL
    ORDER BY cr.namespace, cr.action, wp.last_used DESC
""")
cap_embeddings = {}
for name, emb in cur.fetchall():
    if emb:
        e = np.array(json.loads(emb) if isinstance(emb, str) else emb, dtype=np.float32)
        e = e / max(np.linalg.norm(e), 1e-12)
        cap_embeddings[name] = e
print(f'{len(cap_embeddings)} cap embeddings loaded')

# Build eval set: 20% holdout from each cap (by trace)
np.random.seed(42)
train_intents = {}
test_intents = []

for cap_name, intents in cap_intents.items():
    n = len(intents)
    perm = np.random.permutation(n)
    split = max(1, int(n * 0.8))
    train_intents[cap_name] = [intents[i] for i in perm[:split]]
    for i in perm[split:]:
        test_intents.append((cap_name, intents[i]))

print(f'Train: {sum(len(v) for v in train_intents.values())} intents')
print(f'Test:  {len(test_intents)} intents')

# Score: for each test intent, find the cap with highest cosine similarity
# among all cap SHGAT embeddings. This simulates discover scoring.
cap_names_list = list(cap_embeddings.keys())
cap_matrix = np.stack([cap_embeddings[n] for n in cap_names_list])

def eval_hit1(test_set):
    hit1 = 0
    hit5 = 0
    for target_cap, intent_emb in test_set:
        if target_cap not in cap_embeddings:
            continue
        intent = intent_emb / max(np.linalg.norm(intent_emb), 1e-12)
        scores = cap_matrix @ intent
        try:
            target_idx = cap_names_list.index(target_cap)
        except ValueError:
            continue
        rank = int(np.sum(scores > scores[target_idx])) + 1
        if rank == 1: hit1 += 1
        if rank <= 5: hit5 += 1
    n = len([1 for t, _ in test_set if t in cap_embeddings])
    return hit1, hit5, n

# Baseline (no capping)
h1, h5, n = eval_hit1(test_intents)
print(f'\nBaseline (no cap):  Hit@1={h1}/{n} ({100*h1/n:.1f}%)  Hit@5={h5}/{n} ({100*h5/n:.1f}%)')
print(f'  (This measures SHGAT embedding quality for discover, not GRU training)')
print(f'  The capping affects GRU training data distribution, not SHGAT embeddings directly.')
print(f'  Impact on GRU requires a full training run.')

337 cap embeddings loaded
Train: 1687 intents
Test:  476 intents

Baseline (no cap):  Hit@1=90/476 (18.9%)  Hit@5=142/476 (29.8%)
  (This measures SHGAT embedding quality for discover, not GRU training)
  The capping affects GRU training data distribution, not SHGAT embeddings directly.
  Impact on GRU requires a full training run.


## 6. Training data quality: before vs after capping

In [7]:
# Key question: does the capped training set have better intent diversity?
# Measure: mean pairwise cosine sim within each cap (lower = more diverse)

def measure_diversity(cap_intents_dict, max_per_cap=None):
    if max_per_cap:
        capped_dict, _ = cap_examples_per_target(cap_intents_dict, max_per_cap)
    else:
        capped_dict = cap_intents_dict
    
    total_examples = 0
    weighted_sim = 0
    caps_with_diversity = 0
    
    per_cap_sims = []
    for cap_name, intents in capped_dict.items():
        if len(intents) < 2:
            continue
        embs = np.stack(intents)
        norms = np.linalg.norm(embs, axis=1, keepdims=True)
        normed = embs / np.maximum(norms, 1e-12)
        sim_matrix = normed @ normed.T
        n = len(intents)
        mean_sim = (sim_matrix.sum() - np.trace(sim_matrix)) / (n * (n-1))
        per_cap_sims.append((cap_name, mean_sim, n))
        weighted_sim += mean_sim * n
        total_examples += n
        caps_with_diversity += 1
    
    return {
        'weighted_mean_sim': weighted_sim / total_examples if total_examples > 0 else 0,
        'total_examples': total_examples,
        'caps_measured': caps_with_diversity,
        'per_cap': sorted(per_cap_sims, key=lambda x: -x[2])[:10]  # top by count
    }

print(f'{"Cap":>6s} | {"Examples":>8s} | {"Weighted mean sim":>18s} | {"Caps measured":>13s}')
print('-' * 55)

for max_cap in [None, 100, 50, 30, 20, 10]:
    d = measure_diversity(train_intents, max_cap)
    label = 'NONE' if max_cap is None else str(max_cap)
    print(f'{label:>6s} | {d["total_examples"]:8d} | {d["weighted_mean_sim"]:18.4f} | {d["caps_measured"]:13d}')

print(f'\n(Lower weighted mean sim = more diverse intents per cap)')

   Cap | Examples |  Weighted mean sim | Caps measured
-------------------------------------------------------
  NONE |     1462 |             0.7383 |           106
   100 |      815 |             0.7798 |           106
    50 |      762 |             0.7868 |           106
    30 |      686 |             0.7820 |           106
    20 |      626 |             0.7791 |           106
    10 |      495 |             0.7728 |           106

(Lower weighted mean sim = more diverse intents per cap)


## 7. Per-cap detail: dominant cap before vs after

In [8]:
# Show the dominant cap's intent clustering before/after capping
dominant_cap = cap_counts.most_common(1)[0][0]
dom_intents = train_intents[dominant_cap]
print(f'{dominant_cap}: {len(dom_intents)} train intents')

# Before capping: mean pairwise sim
dom_embs = np.stack(dom_intents)
dom_norms = np.linalg.norm(dom_embs, axis=1, keepdims=True)
dom_normed = dom_embs / np.maximum(dom_norms, 1e-12)
full_sim = dom_normed @ dom_normed.T
n = len(dom_intents)
full_mean = (full_sim.sum() - np.trace(full_sim)) / (n * (n-1))
print(f'  Full ({n}):  mean pairwise sim = {full_mean:.4f}')

# After capping at 30
fps_idx = farthest_point_sample(dom_embs, 30)
fps_normed = dom_normed[fps_idx]
fps_sim = fps_normed @ fps_normed.T
k = len(fps_idx)
fps_mean = (fps_sim.sum() - np.trace(fps_sim)) / (k * (k-1))
print(f'  FPS (30): mean pairwise sim = {fps_mean:.4f}')

# Random 30
rng = np.random.RandomState(42)
rand_idx = rng.choice(n, size=30, replace=False)
rand_normed = dom_normed[rand_idx]
rand_sim = rand_normed @ rand_normed.T
rand_mean = (rand_sim.sum() - np.trace(rand_sim)) / (30 * 29)
print(f'  Random (30): mean pairwise sim = {rand_mean:.4f}')

print(f'\n  FPS keeps more diverse intents: sim {full_mean:.4f} → {fps_mean:.4f} ({"better" if fps_mean < full_mean else "worse"})')
print(f'  vs random:                      sim {full_mean:.4f} → {rand_mean:.4f} ({"better" if rand_mean < full_mean else "worse"})')

db:postgresQuery: 747 train intents
  Full (747):  mean pairwise sim = 0.6822
  FPS (30): mean pairwise sim = 0.6188
  Random (30): mean pairwise sim = 0.6829

  FPS keeps more diverse intents: sim 0.6822 → 0.6188 (better)
  vs random:                      sim 0.6822 → 0.6829 (worse)


## 8. Summary

In [9]:
# Final recommendation
total = len(traces)
dom_count = cap_counts.most_common(1)[0][1]
dom_name = cap_counts.most_common(1)[0][0]

capped_30, stats_30 = cap_examples_per_target(cap_intents, 30)
after_total = sum(len(v) for v in capped_30.values())
top_after = max(len(v) for v in capped_30.values())

print('=' * 60)
print('PER-CAP FREQUENCY CAPPING — SUMMARY')
print('=' * 60)
print(f'''
Problem:
  {dom_name} = {dom_count}/{total} traces ({100*dom_count/total:.1f}%)
  Median cap = {np.median(counts):.0f} traces, mean = {np.mean(counts):.1f}
  Top 2 caps = {sum(c for _, c in cap_counts.most_common(2))}/{total} ({100*sum(c for _, c in cap_counts.most_common(2))/total:.1f}%)

Fix: MAX_PER_CAP=30 with farthest-point sampling
  Before: {total} examples, top cap = {100*dom_count/total:.1f}%
  After:  {after_total} examples, top cap = {100*top_after/after_total:.1f}%
  Dropped: {total - after_total} ({100*(total-after_total)/total:.1f}%) — redundant intents
  FPS preserves intent diversity (lower mean pairwise sim)

Implementation:
  lib/gru/src/data-prep/cap-frequency-cap.ts
  Applied in benchmark-e2e.ts + train-gru-with-caps.ts
  ENV: MAX_PER_CAP=30 (default), MAX_PER_CAP=99999 to disable

Next: Run GRU training with capping to measure Hit@1 impact.
''')

PER-CAP FREQUENCY CAPPING — SUMMARY

Problem:
  db:postgresQuery = 934/2163 traces (43.2%)
  Median cap = 2 traces, mean = 6.5
  Top 2 caps = 1001/2163 (46.3%)

Fix: MAX_PER_CAP=30 with farthest-point sampling
  Before: 2163 examples, top cap = 43.2%
  After:  1135 examples, top cap = 2.6%
  Dropped: 1028 (47.5%) — redundant intents
  FPS preserves intent diversity (lower mean pairwise sim)

Implementation:
  lib/gru/src/data-prep/cap-frequency-cap.ts
  Applied in benchmark-e2e.ts + train-gru-with-caps.ts
  ENV: MAX_PER_CAP=30 (default), MAX_PER_CAP=99999 to disable

Next: Run GRU training with capping to measure Hit@1 impact.



## Post-hoc Update (2026-02-28)

**Verdict**: MAX_PER_CAP=30 was too aggressive (-12pp Tool Hit@1). The frequency capping approach (dropping data) has been **superseded by cap-source sample weighting** in the training loss.

Instead of removing examples from dominant caps, we now assign per-example weights based on the source capability frequency: `weight = sqrt(1 / cap_trace_count)`, mean-normalized. This achieves the same rebalancing with **zero information loss**.

- Cap=30 FPS: 47.5% data dropped, Tool -12pp → **NOT RECOMMENDED**
- Cap-source weighting: 0% data dropped, ratio ~3x → **CURRENT APPROACH**

The `MAX_PER_CAP` default has been changed to 99999 (disabled) in `benchmark-e2e.ts`.
See NB24 for class weight scheme comparison (target-level), and benchmark logs for cap-source results.